app/src/native-notifications.es6

```diff
@@ -35,13 +43,83 @@ class NativeNotifications {
     return false;
   }
 
+  /**
+   * Check if the desktop file exists and parse the desktop file for the icon.
+   *
+   * @param {string} filePath to the desktop file
+   * @returns {string} icon from the desktop file
+   */
+  readIconFromDesktopFile(filePath) {
+    if (fs.existsSync(filePath)) {
+      const ini = require('ini');
+      const content = ini.parse(fs.readFileSync(filePath, 'utf-8'));
+      return content['Desktop Entry']['Icon'];
+    }
+    return null;
+  }
+
+  /**
+   * Get notification icon. Only works on linux, otherwise the Mailspring default icon wil be read
+   * from resources.
+   *
+   * Reading the icon name from the desktop file of Mailspring. If the icon is a name, reads the
+   * icon theme directory for this icon. As the notification only works with PNG files, SVG files
+   * must be converted to PNG
+   *
+   * @returns {string} path to the icon
+   * @private
+   */
+  __getIcon() {
+    if (platform === 'linux') {
+      const desktopBaseDirs = [
+        os.homedir() + '/.local/share/applications/',
+        '/usr/share/applications/',
+      ];
+      const desktopFileNames = ['mailspring.desktop', 'Mailspring.desktop'];
+      // check the applications directories, the user directory has a higher priority
+      for (const baseDir of desktopBaseDirs) {
+        // check multiple spellings
+        for (const fileName of desktopFileNames) {
+          const filePath = path.join(baseDir, fileName);
+          const desktopIcon = this.readIconFromDesktopFile(filePath);
+          if (desktopIcon != null) {
+            if (fs.existsSync(desktopIcon)) {
+              // icon is a file and can be returned
+              return desktopIcon;
+            }
+            // icon is a name and we need to get it from the icon theme
+            const iconPath = getIcon(desktopIcon || 'mailspring', 64, Context.APPLICATIONS, 2);
+            if (iconPath != null) {
+              // only .png icons work with notifications
+              if (path.extname(iconPath) === '.png') {
+                return iconPath;
+              }
+              // try to read it from cache
+              if (ICON_CACHE.hasOwnProperty(iconPath)) {
```

Comment:
Thanks for adding a cache here - I'm a little worried that `getIcon` runs a bunch of sync filesystem + exec code that might be pretty slow on older machines... it'd be nice if the cache prevented that work from happening over and over as well. 

What if we ran `__getIcon` once when the application starts (eg we add  `let resolvedIcon = null` at the top of this file and then set it the first time a notification is needed), and then re-use `resolvedIcon` each time? You'd need to restart Mailspring if you changed your theme, but that might be an acceptable compromise to avoid the overhead of scanning for the icon / parsing ini, etc. each time!

Reply:
I thought about it and even wanted to ask you about which solution you would prefer. But then i saw your typescript PR and was so flashed that i forgot everything else :sweat_smile: